src/js/ripple/amount.js

```diff
@@ -110,170 +120,47 @@ Amount.prototype.abs = function() {
-    var v2  = v._is_negative ? v._value.negate() : v._value;
-    var s   = v1.add(v2);
-
-    result._is_negative = s.compareTo(BigInteger.ZERO) < 0;
-    result._value       = result._is_negative ? s.negate() : s;
-    result._currency    = this._currency;
-    result._issuer      = this._issuer;
-  } else {
-    var v1  = this._is_negative ? this._value.negate() : this._value;
-    var o1  = this._offset;
-    var v2  = v._is_negative ? v._value.negate() : v._value;
-    var o2  = v._offset;
-
-    while (o1 < o2) {
-      v1  = v1.divide(Amount.bi_10);
-      o1  += 1;
-    }
-
-    while (o2 < o1) {
-      v2  = v2.divide(Amount.bi_10);
-      o2  += 1;
-    }
-
-    result = new Amount();
-    result._is_native = false;
-    result._offset = o1;
-    result._value = v1.add(v2);
-    result._is_negative = result._value.compareTo(BigInteger.ZERO) < 0;
-
-    if (result._is_negative) {
-      result._value = result._value.negate();
-    }
-
-    result._currency = this._currency;
-    result._issuer = this._issuer;
+Amount.prototype.add = function(addend) {
+  var addendAmount = Amount.from_json(addend);
 
-    result.canonicalize();
+  if (!this.is_comparable(addendAmount)) {
+    return Amount.NaN();
   }
 
-  return result;
+  return this._copy(this._value.plus(addendAmount._value));
 };
 
-// Result in terms of this currency and issuer.
-Amount.prototype.subtract = function(v) {
+Amount.prototype.subtract = function(subtrahend) {
   // Correctness over speed, less code has less bugs, reuse add code.
-  return this.add(Amount.from_json(v).negate());
+  return this.add(Amount.from_json(subtrahend).negate());
 };
 
-// Result in terms of this' currency and issuer.
 // XXX Diverges from cpp.
-Amount.prototype.multiply = function(v) {
-  var result;
-
-  v = Amount.from_json(v);
-
-  if (this.is_zero()) {
-    result = this;
-  } else if (v.is_zero()) {
-    result = this.clone();
-    result._value = BigInteger.ZERO;
-  } else {
-    var v1 = this._value;
-    var o1 = this._offset;
-    var v2 = v._value;
-    var o2 = v._offset;
-
-    if (this.is_native()) {
-      while (v1.compareTo(Amount.bi_man_min_value) < 0) {
-        v1 = v1.multiply(Amount.bi_10);
-        o1 -= 1;
-      }
-    }
-
-    if (v.is_native()) {
-      while (v2.compareTo(Amount.bi_man_min_value) < 0) {
-        v2 = v2.multiply(Amount.bi_10);
-        o2 -= 1;
-      }
-    }
-
-    result              = new Amount();
-    result._offset      = o1 + o2 + 14;
-    result._value       = v1.multiply(v2).divide(Amount.bi_1e14).add(Amount.bi_7);
-    result._is_native   = this._is_native;
-    result._is_negative = this._is_negative !== v._is_negative;
-    result._currency    = this._currency;
-    result._issuer      = this._issuer;
-
-    result.canonicalize();
-  }
-
-  return result;
+Amount.prototype.multiply = function(multiplicand) {
+  var multiplicandAmount = Amount.from_json(multiplicand);
+  // TODO: probably should just multiply by multiplicandAmount._value
```

Comment:
`divide` seems to go to some troubles to check validity of the operands, yet this doesn't ?